# Finding similar company name and auto matching them -  editted pr450

This program will use NLP and ML technique to match similar company names. Matching form common words like "LTD" and "COMPANY" will be discounted autometically in the algorithm.

Library used:
* pandas
* fuzzywuzzy

## Load the table in pandas DataFrame

Data later deleted but found on http://download.companieshouse.gov.uk/en_output.html *(the version shown here is snapshot of June 2020)*

In [15]:
import pandas as pd
pd.set_option('display.max_columns', 1000)
import os
os.chdir("/home/puria-radmard/Repos/VectorAI2020/fuzzy-match-company-name")
print("CP 1")
df = pd.read_csv("BasicCompanyDataAsOneFile-2020-06-01.csv", sep='\t')
print("CP 2")
df.head()

# Open up the csv which we will work with. 2.2GB so later deleted to allow commits

CP 1


In [1]:
df.columns

# All features by which companies are described wtih in the dataset

NameError: name 'df' is not defined

In [2]:
df['RegAddress.PostTown'].value_counts().head(30)

NameError: name 'df' is not defined

## Frequency of words
Since we have lots of companies, we will only use companies in Cambridge as an example.

First we find the 30 most common words in all company names. As we will be expecting them to be repeating a lot even in companies that is not the same, we cannot match company names using them. The way we do it is we will deduct the matching score of a pair if any keywords is present in the names.

In [4]:
from collections import Counter
all_names = df['CompanyName'][df['RegAddress.PostTown']=='CAMBRIDGE'].unique()
names_freq = Counter()
for name in all_names:
    names_freq.update(str(name).split(" "))
key_words = [word for (word,_) in names_freq.most_common(30)]
print(key_words)

['LIMITED', 'LTD', 'CAMBRIDGE', 'SERVICES', 'MANAGEMENT', '&', 'COMPANY', 'THE', 'CONSULTING', 'LTD.', 'SOLUTIONS', 'AND', 'PROPERTY', 'UK', 'LLP', '(CAMBRIDGE)', 'CONSULTANCY', 'GROUP', 'HOLDINGS', 'CONSULTANTS', 'ASSOCIATES', 'COMPOSITES', 'ENGINEERING', 'DEVELOPMENTS', 'INTERNATIONAL', 'OF', 'DESIGN', 'TECHNOLOGY', 'PROPERTIES', '(UK)']


In [5]:
len(all_names)

15889

## Matching by Grouping
Then we group the names by their 1st character. As the list is too long, it will take forever to match them all at once (15889 x 15889 pairs to consider). The work around is to match them by groups, assuming if the names are not matched at the 1st character, it is unlikely that they are the same name. 

In [6]:
all_main_name = pd.DataFrame(columns=['sort_gp','names','alias','score'])
all_names.sort()
all_main_name['names'] = all_names
all_main_name['sort_gp'] = all_main_name['names'].apply(lambda x: x[0])

## Fuzzy Matching
Here for each group, we use `fuzzywuzzy.token_sort_ratio` to matching the names. Different form the basic `fuzzywuzzy.ratio` which use Levenshtein Distance to calculate the differences, it allow the token (words) in a name to swap order and still give a 'perfect' match. (ref: https://github.com/seatgeek/fuzzywuzzy)

In [7]:
from fuzzywuzzy import fuzz

all_sort_gp = all_main_name['sort_gp'].unique()

def no_key_word(name):
    """check if the name contain the keywords in travel company"""
    output = True
    for key in key_words:
        if key in name:
            output = False
    return output

for sortgp in all_sort_gp:
    this_gp = all_main_name.groupby(['sort_gp']).get_group(sortgp)
    gp_start = this_gp.index.min()
    gp_end = this_gp.index.max()
    for i in range(gp_start,gp_end+1):
    
        # if self has not got alias, asign to be alias of itself
        if pd.isna(all_main_name['alias'].iloc[i]):
            all_main_name['alias'].iloc[i] = all_main_name['names'].iloc[i]
            all_main_name['score'].iloc[i] = 100
        
        # if the following has not got alias and fuzzy match, asign to be alias of this one
        for j in range(i+1,gp_end+1):
            if pd.isna(all_main_name['alias'].iloc[j]):
                fuzz_socre = fuzz.token_sort_ratio(all_main_name['names'].iloc[i],all_main_name['names'].iloc[j])
                if not no_key_word(all_main_name['names'].iloc[j]):
                    fuzz_socre -= 10
                if (fuzz_socre > 85):
                    all_main_name['alias'].iloc[j] = all_main_name['alias'].iloc[i]
                    all_main_name['score'].iloc[j] = fuzz_socre
                    
        if i % (len(all_names)//10) == 0:
            print("progress: %.2f" % (100*i/len(all_names)) + "%")
                
all_main_name.to_csv('company_in_cambridge.csv')

progress: 0.00%
progress: 9.99%
progress: 19.99%
progress: 29.98%
progress: 39.98%
progress: 49.97%
progress: 59.97%
progress: 69.96%
progress: 79.95%
progress: 89.95%
progress: 99.94%


In [8]:
all_main_name[(all_main_name['names']!=all_main_name['alias']) & (all_main_name['alias'].notna())]

sort_gp                                              names  \
761         A                                     AMADEUS EII LP   
762         A                                    AMADEUS EIII LP   
763         A                                      AMADEUS HI LP   
766         A                                     AMADEUS II 'A'   
767         A                                     AMADEUS II 'B'   
768         A                                     AMADEUS II 'C'   
769         A                                        AMADEUS III   
773         A                   AMADEUS IV EARLY STAGE FUND B LP   
776         A                                      AMADEUS JI LP   
777         A                                      AMADEUS LI LP   
783         A                                      AMADEUS TI LP   
937         A                            ANP ENGINEERING LIMITED   
1361        A                                      AVITO LIMITED   
1607        B                            BCK INVESTMENTS LIMITED   
1849        B                                    BIRDY FILMS LTD   
1971        B               BLUESTONE CAPITAL MANAGEMENT LIMITED   
1976        B           BLUESTONE MORTGAGE FINANCE NO. 4 LIMITED   
1977        B            BLUESTONE MORTGAGE FINANCE NO.2 LIMITED   
2698        C               CAMBRIDGE DESIGN PARTNERSHIP LIMITED   
2839        C                          CAMBRIDGE H C 102 LIMITED   
2840        C                          CAMBRIDGE H C 129 LIMITED   
2911        C   CAMBRIDGE INTERNATIONAL EDUCATION CENTRE LIMITED   
3142        C         CAMBRIDGE RESEARCH AND ENGINEERING LIMITED   
3184        C   CAMBRIDGE SEMINARS INTERNATIONAL COLLEGE LIMITED   
3566        C               CASA MIA DEVELOPMENTS (NO 3) LIMITED   
4159        C                COLLEGE COURT (CAMBRIDGE B) LIMITED   
4944        D                       DGP ENGINEERING SERVICES LTD   
5182        D                  DS ELECTRICAL (CAMBRIDGE) LIMITED   
6220        F                       FRAZER STANNARD PROPERTY LTD   
6407        G           GATEWAY (218) MANAGEMENT COMPANY LIMITED   
6517        G                                       GFB AERO LTD   
6951        H                HAMMER AND TONG PRODUCTIONS LIMITED   
7354        H                           HPG DEVELOPMENTS LIMITED   
7811        I                              IQ CAPITAL FUND II LP   
7812        I                           IQ CAPITAL FUND III C LP   
7813        I                             IQ CAPITAL FUND III LP   
7817        I                     IQ CAPITAL PARTNERS GP III LLP   
8420        K  KEW BRIDGE WEST 2 RESIDENTS MANAGEMENT COMPANY...   
8421        K  KEW BRIDGE WEST 3 RESIDENTS MANAGEMENT COMPANY...   
9772        M                           MILLERS PARTNERS LIMITED   
11225       P                        PIGEON HUNSTANTON 2 LIMITED   
11711       Q          QQ2 RESIDENTS PROPERTY MANAGEMENT LIMITED   
12129       R  RIDGEMOND PARK (BLOCK 4) MANAGEMENT COMPANY LI...   
12156       R                           RINDL CONSULTING LIMITED   
12523       S                              SANDFORD AVIATION LTD   
12807       S                    SHAWS OF CAMBRIDGE (EH) LIMITED   
12902       S   SILKWORKS RESIDENTS MANAGEMENT COMPANY 3 LIMITED   
13499       S                       STOW HEALTHCARE (FP) LIMITED   
13865       T                                       TCOM LIMITED   
14116       T                               THE CUWBC FOUNDATION   
14149       T            THE EXETER TRUSTEE COMPANY NO.2 LIMITED   
14343       T                THE PROCUREMENT PARTNERSHIP LIMITED   
14702       T                         TRANQUILLITY PARKS LIMITED   
14816       T                           TSANG CONSULTING LIMITED   
15340       W                      WATER LANE 3 NOMINEES LIMITED   
15349       W         WATERBEACH NO.1 MANAGEMENT COMPANY LIMITED   
15667       W                 WRENBRIDGE (CHEDDARS LANE) LIMITED   

                                                   alias  sco

The result is saved in a csv file locally for future inspection and further experimentation. Inspecting the result, the matches consisted of 3 groups:

1. they are usually differ in spelling by 1 character: missing an 'L' or 'I' or 'S'
2. highly similar names: 'No.3' instead of 'No.2' or 'EB' instread of 'EH'
3. fairly similar names: 'HAMMER AND THONGS PRODUCTIONS LIMITED' and 'HAMMER AND TONG PRODUCTIONS LIMITED'

For type 1 and 2 matches it could be the same company, the diffeernce in names could be an intentional alteration or simply a typo. But it is not likely the same company for type 3 matched, it seems more like a coincidnce. 

To further confirm, manual work need to be done but this program saves a lot of manual work hours.

In [9]:
all_main_name[(all_main_name['names']!=all_main_name['alias']) & (all_main_name['alias'].notna())].shape[0]

57

In [10]:
len(all_main_name['alias'].unique())

15832

By applying the fuzzy matching, 57 names are caught similar to another name, which is less then 1% of the total. By using this program names that need checking drastically reduce form 15889 total to only 57.